File: algorithms/0112.path-sum/solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} sum
# @return {Boolean}
def has_path_sum(root, sum)
  return false if root.nil?

  if root.val == sum
    if root.left.nil? && root.right.nil?
      return true
    end
  end

  has_path_sum(root.left, sum - root.val) || has_path_sum(root.right, sum - root.val)
end

File: algorithms/0112.path-sum/solution.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, sum: i32) -> bool {
        if let Some(node) = &root {
            let node = node.borrow();
            if node.val == sum {
                if node.left.is_none() && node.right.is_none() {
                    return true;
                }
            }
            return Self::has_path_sum(node.left.clone(), sum - node.val)
                || Self::has_path_sum(node.right.clone(), sum - node.val);
        }

        false
    }
}
